### PR TITLE
Add virtual machine summary

### PR DESCRIPTION
**Pull request by coliver**
This commit also adds VirtualMachineConfigSummary and VirtualMachineStorageSummary since they are linked off of VirtualMachineSummary

### CODE REVIEW

File: lib/fog/vsphere/models/compute/server.rb [hunk 1]
```diff
@@ -17,46 +17,47 @@ class Server < Fog::Compute::Server
...
+        attribute :power_state,   :aliases => 'power'
+        attribute :public_ip_address, :aliases => 'ipaddress'
+        attribute :resource_pool
+        attribute :scsi_controller # this is the first scsi controller. Right now no more of them can be used.
+        attribute :summary # VirtualMachineSummary
+        attribute :tools_state,   :aliases => 'tools'
+        attribute :tools_version
+        # UUID may be the same from VM to VM if the user does not select (I copied it)
+        attribute :uuid
+        attribute :volumes
```
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/server.rb]: Did you just sorted them? :heart:

Comment (jrgarcia) [inline lib/fog/vsphere/models/compute/server.rb]: I will never complain about sorted attributes :+1:

File: .gitignore [hunk 2]
```diff
@@ -20,4 +20,6 @@ tmp
 *.o
 *.a
 mkmf.log
-gemfiles/*.lock
\ No newline at end of file
+gemfiles/*.lock
+*.ruby-*
```
Comment (plribeiro3000) [inline .gitignore]: Why did you add this to `gitignore`?
I think they are good to have because all developers will be on the same version and thus discussion wont be affected by different behaviours through ruby versions.
And we can always trust `travis` to test all versions for us. =)

File: .gitignore [hunk 3]
```diff
@@ -20,4 +20,6 @@ tmp
 *.o
 *.a
 mkmf.log
-gemfiles/*.lock
\ No newline at end of file
+gemfiles/*.lock
```
Comment (jrgarcia) [inline .gitignore]: :+1: Thanks for this! Can't believe I missed it...

File: lib/fog/vsphere/models/compute/server.rb [hunk 4]
```diff
@@ -319,6 +320,13 @@ def initialize_scsi_controller
             Fog::Compute::Vsphere::SCSIController.new(self.attributes[:scsi_controller])
           end
         end
+
+        def initialize_virtual_machine_summary
+          if attributes[:summary] && attributes[:summary].props
+            self.attributes[:summary] = Fog::Compute::Vsphere::VirtualMachineSummary.new(self.attributes[:summary].props)
```
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/server.rb]: Is this code initializing an association?

If so what do you think of using `fog` DSL for it?

See examples at this [test](https://github.com/fog/fog-core/blob/master/spec/fog_attribute_spec.rb#L52-L57).

File: lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        attribute :uuid
+        attribute :vmPathName
+
+        def initialize(attributes={})
+          super
+          #TODO: Other objects get initialized here
+          #TODO: For now they will be their rbvmomi counterparts
+          #TODO: self.ftInfo = FaultToleranceConfigInfo.new(attributes[:ftInfo].props) if attributes[:ftInfo] && attributes[:ftInfo].props
+          #TODO: self.product = VAppProductInfo.new(attributes[:product].props) if attributes[:product] && attributes[:product].props
+        end
```
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb]: Could you remove this code if it is not doing anything yet?
I know you will add something but i feel if we don't need it yet better not add it since it will add one extra loop when evaluating this class.
And there is also the fact that if you are thinking about associations `Fog` have a builtin DSL for it.

File: lib/fog/vsphere/models/compute/virtual_machine_summary.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        attribute :runtime        #VirtualMachineRuntimeInfo
+        attribute :storage        #VirtualMachineStorageSummary
+        attribute :vm             #ManagedObjectReference to a VirtualMachine
+
+        def initialize(attributes={})
+          super
+          #TODO: Other objects should get initialized here
+          #TODO: For now they will be their rbvmomi equivalents by default
+          self.storage = VirtualMachineStorageSummary.new(attributes[:storage].props) if attributes[:storage] && attributes[:storage].props
+          self.config = VirtualMachineConfigSummary.new(attributes[:config].props) if attributes[:config] && attributes[:config].props
```
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/virtual_machine_summary.rb]: Same here for the association DSL usage.

File: .gitignore (hunk 2 above)
Comment (coliver) [inline .gitignore]: I agree that having ruby-version and ruby-gemset in the project is good, but they were not in there already so I was sticking with what was in there.  Let me know if I should change this.

File: lib/fog/vsphere/models/compute/server.rb (hunk 4 above)
Comment (coliver) [inline lib/fog/vsphere/models/compute/server.rb]: So, instead of this method server should has_one virtual_machine_summary?

File: lib/fog/vsphere/models/compute/server.rb (hunk 1 above)
Comment (coliver) [inline lib/fog/vsphere/models/compute/server.rb]: Yes. I use a plugin for RubyMine (I recommend it) called, surprisingly, sort lines that makes it very easy. :)

File: lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb (hunk 5 above)
Comment (coliver) [inline lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb]: With the exception of the two objects I haven't defined yet, ftInfo and product, all of the other attributes are basic types (strings, bools, and ints).  I figured those was useful. If you want to wait until those are defined I can remove this.

File: .gitignore (hunk 2 above)
Comment (coliver) [inline .gitignore]: Also, today I learned there is a ~/.gitignore.  I should have never pushed this file... :/ sorry.

Comment (plribeiro3000) [inline .gitignore]: Thats not a problem. =)
We do need a `.gitignore` file here anyway. :+1:

File: lib/fog/vsphere/models/compute/server.rb (hunk 4 above)
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/server.rb]: Yep. This will get the attributes and save in another instance variable inside the object
and when calling the getter it will get those attributes and initialize an object for you out of the box.

File: lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb (hunk 5 above)
Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/virtual_machine_config_summary.rb]: Got it, but i believe you can solve this code here using the DSL for associations as well.

File: lib/fog/vsphere/models/compute/server.rb (hunk 4 above)
Comment (coliver) [inline lib/fog/vsphere/models/compute/server.rb]: I tried playing with this last week but was unable to get it going.  Other than that test, is there an example in another fog module I could look at to get it working?  I scanned through aws and google but didn't see it being used.

Comment (plribeiro3000) [inline lib/fog/vsphere/models/compute/server.rb]: Yeah, so that feature was added because of `fog-xenserver`. Take a look at its models [here](https://github.com/fog/fog-xenserver/tree/master/lib/fog/compute/xen_server/models). None of the other providers did started using this feature yet but once they do lots of code won't be needed anymore.

**Comment by jrgarcia** [comment]
@coliver Sorry, I know this is a dated PR, but are you planning on making any of the changes suggested (or resolving the conflicts with the latest code)?

**Comment by coliver** [comment]
I couldn't get that dsl working, so I ended up using https://github.com/rlane/rbvmomi to get what I needed for the project I was working on, so, no, sorry. I'll retract the pull request.

**Comment by jrgarcia** [comment]
No problem! I'm sorry it didn't work out for you. Thanks for closing this!
